src/Columns/ColumnDynamic.cpp

```diff
@@ -1021,54 +1021,63 @@ ColumnPtr ColumnDynamic::compress(bool force_compression) const
     {
-        variants[i]->updateCheckpoint(*nested[i]);
     }
 
-    /// If column has new variants since last checkpoint create checkpoints for them.
-    if (old_size < variants.size())
+    std::unordered_map<String, ColumnCheckpointPtr> variants_checkpoints;
+};
+
+}
+
+ColumnCheckpointPtr ColumnDynamic::getCheckpoint() const
+{
+    std::unordered_map<String, ColumnCheckpointPtr> variants_checkpoints;
+    for (const auto & [name, discr] : variant_info.variant_name_to_discriminator)
+        variants_checkpoints[name] = variant_column_ptr->getVariantByGlobalDiscriminator(discr).getCheckpoint();
+    return std::make_shared<DynamicColumnCheckpoint>(size(), variants_checkpoints);
+}
+
+void ColumnDynamic::updateCheckpoint(ColumnCheckpoint & checkpoint) const
+{
+    auto & variants_checkpoints = assert_cast<DynamicColumnCheckpoint &>(checkpoint).variants_checkpoints;
+
+    for (const auto & [name, discr] : variant_info.variant_name_to_discriminator)
     {
-        nested.resize(variants.size());
-        for (size_t i = old_size; i < variants.size(); ++i)
-            nested[i] = variants[i]->getCheckpoint();
+        auto it = variants_checkpoints.find(name);
+        /// If column has new variants since last checkpoint create checkpoints for them.
+        if (it == variants_checkpoints.end())
+            variants_checkpoints.emplace(name, variant_column_ptr->getVariantByGlobalDiscriminator(discr).getCheckpoint());
+        else
+            variant_column_ptr->getVariantByGlobalDiscriminator(discr).updateCheckpoint(*it->second);
     }
 
     checkpoint.size = size();
 }
 
 void ColumnDynamic::rollback(const ColumnCheckpoint & checkpoint)
 {
-    const auto & nested = assert_cast<const ColumnCheckpointWithMultipleNested &>(checkpoint).nested;
-    chassert(nested.size() <= variant_column_ptr->getNumVariants());
-
-    /// The structure hasn't changed, so we can use generic rollback of Variant column
-    if (nested.size() == variant_column_ptr->getNumVariants())
-    {
-        variant_column_ptr->rollback(checkpoint);
-        return;
-    }
+    const auto & variants_checkpoints = assert_cast<const DynamicColumnCheckpoint &>(checkpoint).variants_checkpoints;
 
     /// Manually rollback internals of Variant column
     variant_column_ptr->getOffsets().resize_assume_reserved(checkpoint.size);
     variant_column_ptr->getLocalDiscriminators().resize_assume_reserved(checkpoint.size);
 
-    auto & variants = variant_column_ptr->getVariants();
-    for (size_t i = 0; i < nested.size(); ++i)
-        variants[i]->rollback(*nested[i]);
-
-    /// Keep the structure of variant as is but rollback
-    /// to 0 variants that are not in the checkpoint.
-    for (size_t i = nested.size(); i < variants.size(); ++i)
-        variants[i] = variants[i]->cloneEmpty();
+    for (const auto & [name, discr] : variant_info.variant_name_to_discriminator)
+    {
+        auto it = variants_checkpoints.find(name);
```

Comment:
I guess the issue arose because "after extension the order of variant types (and so their discriminators) can change" (from the inline docs of column variant), right?

Reply:
Not exactly. The extension of a Variant happens in a way to preserve local order of columns to avoid discriminators rewrite (so global order of variants is changing, but local in the ColumnVariant is not).
But there is a problem that in `SerializationMap` in case of an error during parsing we use `column->cut()`:
https://github.com/ClickHouse/ClickHouse/blob/090cb5a4fddd5f8202689bb0fcb6108f3c5314d9/src/DataTypes/Serializations/SerializationMap.cpp#L208
And there we clone empty column and do `insertRangeFrom`. And in this case local order will be changed because when we clone empty Variant column we use global order as local (because we don't have discriminators and nothing is needed to rewrite)